src/balancer-strategy-manager/BalancerV2WeightedPoolStrategyManager.sol

```diff
@@ -0,0 +1,233 @@
+   * @param _guardian The guardian of the contract.
+   * @param _hypernative The Hypernative service address.
+   */
+  constructor(
+    address _vault,
+    TokenConfig[] memory _tokenConfig,
+    address _owner,
+    address _guardian,
+    address _hypernative
+  ) {
+    VAULT = IBalancerVault(_vault);
+
+    for (uint256 i = 0; i < _tokenConfig.length; ) {
+      tokenProvider[_tokenConfig[i].token] = _tokenConfig[i].provider;
+
+      emit TokenProviderUpdated(_tokenConfig[i].token, address(0), _tokenConfig[i].provider);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    _transferOwnership(_owner);
+    _updateGuardian(_guardian);
+    HYPERNATIVE = _hypernative;
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function setTokenProvider(address _token, address _provider) external onlyOwner {
+    address oldProvider = tokenProvider[_token];
+    tokenProvider[_token] = _provider;
+
+    emit TokenProviderUpdated(_token, oldProvider, _provider);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function deposit(bytes32 _poolId, uint256[] calldata _tokenAmounts) external onlyOwnerOrGuardian {
+    address[] memory assets;
+    (assets, , ) = VAULT.getPoolTokens(_poolId);
+    uint256 tokenCount = assets.length;
+    (address poolAddress, ) = VAULT.getPool(_poolId);
+
+    if (_tokenAmounts.length != tokenCount) {
+      revert TokenCountMismatch();
+    }
+
+    for (uint256 i = 0; i < tokenCount; ) {
+      uint256 currentBalance = IERC20(assets[i]).balanceOf(address(this));
+
+      if (_tokenAmounts[i] > currentBalance) {
+        revert InsufficientToken(assets[i], currentBalance);
+      }
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    for (uint256 i = 0; i < tokenCount; ) {
+      IERC20(assets[i]).safeIncreaseAllowance(address(VAULT), _tokenAmounts[i]);
+
+      unchecked {
+        ++i;
+      }
+    }
+
+    bytes memory userData = abi.encode(
+      WeightedPoolUserData.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+      _tokenAmounts,
+      0
+    );
+
+    IBalancerVault.JoinPoolRequest memory request = IBalancerVault.JoinPoolRequest({
+      assets: assets,
+      maxAmountsIn: _tokenAmounts,
+      userData: userData,
+      fromInternalBalance: false
+    });
+
+    uint256 bptAmountBefore = IERC20(poolAddress).balanceOf(address(this));
+    VAULT.joinPool(_poolId, address(this), address(this), request);
+    uint256 bptAmountAfter = IERC20(poolAddress).balanceOf(address(this));
+
+    emit TokenDeposit(_msgSender(), _poolId, _tokenAmounts, bptAmountAfter - bptAmountBefore);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function withdraw(
+    bytes32 _poolId,
+    uint256 bpt
+  ) external onlyOwnerOrGuardian returns (uint256[] memory) {
+    return _withdraw(_poolId, bpt);
+  }
+
+  /// @inheritdoc IBalancerStrategyManager
+  function emergencyWithdraw(bytes32 _poolId) external onlyWithdrawable returns (uint256[] memory) {
+    (address poolAddress, ) = VAULT.getPool(_poolId);
```

Comment:
Should we have a multi emergency withdraw instead of a single one? We don't want to have to send 10 Txs if there are multiple pools compromised